service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/INotificationManagementUseCase.kt

```diff
@@ -0,0 +1,56 @@
+package org.thechance.service_notification.domain.usecases
+
+import org.koin.core.annotation.Single
+import org.thechance.service_notification.domain.entity.InternalServerErrorException
+import org.thechance.service_notification.domain.entity.Notification
+import org.thechance.service_notification.domain.gateway.IDatabaseGateway
+import org.thechance.service_notification.domain.gateway.IPushNotificationGateway
+import org.thechance.service_notification.endpoints.TOPIC_NOT_EXISTS
+
+interface INotificationManagementUseCase {
+    suspend fun sendNotificationToUser(userId: String, title: String, body: String): Boolean
+
+    suspend fun sendNotificationToTopic(topic: String, title: String, body: String): Boolean
+
+    suspend fun getNotificationHistory(page: Int, limit: Int): List<Notification>
+}
+
+@Single
+class NotificationManagementUseCase(
+    private val pushNotificationGateway: IPushNotificationGateway,
+    private val databaseGateway: IDatabaseGateway,
+) : INotificationManagementUseCase {
+
+    override suspend fun sendNotificationToUser(userId: String, title: String, body: String): Boolean {
+        val tokens = databaseGateway.getUserTokens(userId)
+        return pushNotificationGateway.sendNotification(tokens, title, body).also {
+            databaseGateway.addNotificationToHistory(
+                Notification(
+                    title = title,
+                    body = body,
+                    date = System.currentTimeMillis(),
+                    userId = userId
+                )
+            )
+        }
+    }
+
+    override suspend fun sendNotificationToTopic(topic: String, title: String, body: String): Boolean {
+        if (!databaseGateway.isTopicAlreadyExists(topic)) throw InternalServerErrorException(TOPIC_NOT_EXISTS)
```

Comment:
Throw Not found exception, not an internal server error if the resource doesn't exist